source/MaterialXTest/ShaderValid.cpp

```diff
@@ -303,22 +309,34 @@ static mx::GlslValidatorPtr createValidator(bool& orthographicView, const std::s
 //
 // Outputs error log if validation fails
 //
-static void runValidation(const std::string& outputPath, const std::string& shaderName, mx::ElementPtr element,
-                          mx::GlslValidatorPtr validator, mx::ShaderGeneratorPtr shaderGenerator,
-                          bool orthographicView, mx::DocumentPtr doc, std::ostream& log, bool outputMtlxDoc=false)
+static void runValidation(const std::string& shaderName, mx::ElementPtr element, mx::GlslValidator& validator,
+                          mx::ShaderGenerator& shaderGenerator, bool orthographicView, mx::DocumentPtr doc,
+                          std::ostream& log, bool outputMtlxDoc=true, const std::string& outputPath=".")
 {
     mx::SgOptions options;
 
     if(element && doc)
     {
         log << "------------ Run validation with element: " << element->getName() << std::endl;
 
-        std::string shaderPath =  mx::FilePath(outputPath) / shaderName;
-        mx::ShaderPtr shader = shaderGenerator->generate(shaderName, element, options);
-        mx::HwShaderPtr hwShader = std::dynamic_pointer_cast<mx::HwShader>(shader);
-        REQUIRE(hwShader != nullptr);
-        REQUIRE(hwShader->getSourceCode(mx::HwShader::PIXEL_STAGE).length() > 0);
-        REQUIRE(hwShader->getSourceCode(mx::HwShader::VERTEX_STAGE).length() > 0);
+        std::string shaderPath;
+        // Note: mkdir will fail if the directory already exists which is ok.
+#if defined(_WIN32)
+        _mkdir(outputPath.c_str());
+#else
+        mkdir(outputFolderPath.c_str(), 0777);
+#endif
+        shaderPath = mx::FilePath(outputPath) / mx::FilePath(shaderName);
+
+        mx::HwShaderPtr hwShader = std::static_pointer_cast<mx::HwShader>(shaderGenerator.generate(shaderName, element, options));
```

Comment:
I still need a cast here because GlslShaderGenerator also generates a mx::ShaderPtr and not a mx::HwShaderPtr.

Reply:
I don't understand why? If it's a GlslShaderGenerator then you already know it will generate a GlslShader which is a HwShader (even if it's stored in a ShaderPtr), so you don't need to cast it to check that. The only method called on it below is getSourceCode() which is available from the ShaderPtr.

Reply:
The reason is that there is no vertex stage vs pixel stage static in the base class. So unsure how you would properly get the source for both stages. Even if we wanted to iterate NUM_STAGES I believe it would return 1 vs 2 (in this case).

Maybe it's easier and also better long term to use named stages vs numeric indexes into an array, especially as the value of the statics changes during class derivation.
 
I assume we don't want to push all stage indexes to the base class as this means all future stages would need to be in the base class.

Reply:
But the shader instance is NOT just a base class instance. We know for a fact that the ShaderPtr is storing a GlslShader instance here, which we know contains both shader stages. So accessing them using these indexes is perfectly fine without any pointer cast.

BTW, my initial suggestion was to change the passed in ShaderGenerator into a GlslShaderGenerator. Because this function only works for GLSL. Maybe that will clarify what I mean, and why the shader is always a GlslShader here? :)

Reply:
Your right, I don't need a HwShader, changing things to just use a shader, hope that's sufficient.